Add the capability to install the `geoipupdate` tool

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This pull request adds functionality to install the MaxMind [geoipupdate] tool. It will also set up a systemd service and timer to run the tool twice a week (by default) to provide automatic update functionality.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

On systems that have a MaxMind GeoIP database added they typically will not have updates for that database until a new AMI is built which leads to outdated data. Since MaxMind provides the [geoipupdate] tool it makes sense to use this to provide a convenient way to update any GeoIP databases either manually or automatically.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

Automated tests pass.
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

[geoipupdate]: https://github.com/maxmind/geoipupdate

mcdonnnj committed:
Add the ability to install the geoipupdate tool

MaxMind provides the geoipupdate tool (https://github.com/maxmind/geoipupdate)
to allow convenient updating of MaxMind GeoIP2 databases. Since this
role installs those databases it makes sense to offer installation of a
tool to manage updating them.

mcdonnnj committed:
Test installation of the geoipupdate tool

Add a new molecule scenario to test installation of the geoipupdate
tool.

mcdonnnj committed:
Add functionality to automatically update GeoIP databases

This is achieved using a SystemD service and timer to run the
geoipupdate tool twice a week. The service and timer and heavily
based on the ones used in the Debian distribution of the geoipupdate
tool:
https://salsa.debian.org/debian/geoipupdate/-/blob/debian/7.0.1-1/debian/service
https://salsa.debian.org/debian/geoipupdate/-/blob/debian/7.0.1-1/debian/timer

mcdonnnj committed:
Ensure the tzdata package is installed on Ubuntu Jammy

This package would be installed on an actual Ubuntu Jammy system per
the manifests:
https://releases.ubuntu.com/jammy/ubuntu-22.04.4-desktop-amd64.manifest
https://releases.ubuntu.com/jammy/ubuntu-22.04.4-live-server-amd64.manifest

However, it is absent on our Docker image and thus we are missing
timezone data. This causes the geoipupdate timer we add to fail when
it is loaded. I verified this was the problem by logging into the
container and when I ran `timedatectl list-timezones` the only result
was "UTC". Once I installed the tzdata package the same command
resulted in an expected list of timezones.

mcdonnnj committed:
Allow some configuration of the geoipupdate SystemD timer

It would be convenient to specify when and with what delay the tool
should be run.

jsf9k commented on README.md:
```diff
@@ -4,7 +4,9 @@
-[MaxMind GeoIP2 database](https://www.maxmind.com/en/geoip2-databases).
+[MaxMind GeoIP2 database](https://www.maxmind.com/en/geoip2-databases). Additionally,
+it can install the [MaxMind `geoipupdate` tool](https://github.com/maxmind/geoipupdate)
+and add a SystemD service and timer to run the tool at regular intervals.
```
["systemd" is all lowercase](https://systemd.io/), I believe.  (I know I have done this incorrectly in the past, but that is no reason to continue my mistakes.)
```suggestion
and add a systemd service and timer to run the tool at regular intervals.
```

mcdonnnj replied:
Yep I was bamboozled somewhere along the way it seems. The [old website has a section specifically around the spelling of systemd](https://www.freedesktop.org/wiki/Software/systemd/#spelling):

> Yes, it is written systemd, not system D or System D, or even SystemD. And it isn't system d either. Why? Because it's a system daemon, and under Unix/Linux those are in lower case, and get suffixed with a lower case d. And since systemd manages the system, it's called systemd. It's that simple. But then again, if all that appears too simple to you, call it (but never spell it!) System Five Hundred since D is the roman numeral for 500 (this also clarifies the relation to System V, right?). The only situation where we find it OK to use an uppercase letter in the name (but don't like it either) is if you start a sentence with systemd. On high holidays you may also spell it sÿstëmd. But then again, [Système D](https://secure.wikimedia.org/wikipedia/en/wiki/System_D) is not an acceptable spelling and something completely different (though kinda fitting).

jsf9k commented on README.md:
```diff
@@ -64,6 +66,12 @@ None.
 |----------|-------------|---------|----------|
 | geoip2\_database\_directory | The directory in which to store the database files. | `/usr/local/share/GeoIP/` | No |
+| geoip2\_geoipupdate\_auto\_update | Whether to configure automatic updates when `geoipupdate` is installed. | `true` | No |
+| geoip2\_geoipupdate\_service\_name | The name to use for the `geoipupdate` SystemD service and timer. | `geoipupdate` | No |
```
```suggestion
| geoip2\_geoipupdate\_service\_name | The name to use for the `geoipupdate` systemd service and timer. | `geoipupdate` | No |
```

jsf9k commented on README.md:
```diff
@@ -64,6 +66,12 @@ None.
 | geoip2\_database\_directory | The directory in which to store the database files. | `/usr/local/share/GeoIP/` | No |
+| geoip2\_geoipupdate\_auto\_update | Whether to configure automatic updates when `geoipupdate` is installed. | `true` | No |
+| geoip2\_geoipupdate\_service\_name | The name to use for the `geoipupdate` SystemD service and timer. | `geoipupdate` | No |
+| geoip2\_geoipupdate\_service\_timer\_on\_calendar | The calendar expression to use for the `geoipupdate` SystemD timer. | `Wed,Sat America/New_York` | No |
```
```suggestion
| geoip2\_geoipupdate\_service\_timer\_on\_calendar | The calendar expression to use for the `geoipupdate` systemd timer. | `Wed,Sat America/New_York` | No |
```

jsf9k commented on README.md:
```diff
@@ -64,6 +66,12 @@ None.
+| geoip2\_geoipupdate\_auto\_update | Whether to configure automatic updates when `geoipupdate` is installed. | `true` | No |
+| geoip2\_geoipupdate\_service\_name | The name to use for the `geoipupdate` SystemD service and timer. | `geoipupdate` | No |
+| geoip2\_geoipupdate\_service\_timer\_on\_calendar | The calendar expression to use for the `geoipupdate` SystemD timer. | `Wed,Sat America/New_York` | No |
+| geoip2\_geoipupdate\_service\_timer\_randomized\_delay\_sec | The randomized delay in seconds to use for the `geoipupdate` SystemD timer. | `3h` | No |
```
```suggestion
| geoip2\_geoipupdate\_service\_timer\_randomized\_delay\_sec | The randomized delay in seconds to use for the `geoipupdate` systemd timer. | `3h` | No |
```

jsf9k commented on defaults/main.yml:
```diff
@@ -10,3 +10,26 @@ geoip2_maxmind_editions:
+# it is treated as a string.
+geoip2_geoipupdate_version: "7.0.1"
+
+# Indicate if a SystemD service and timer should be added to automatically
```
```suggestion
# Indicate if a systemd service and timer should be added to automatically
```

jsf9k commented on defaults/main.yml:
```diff
@@ -10,3 +10,26 @@ geoip2_maxmind_editions:
+# Indicate if a SystemD service and timer should be added to automatically
+# update the GeoIP database files when the geoipupdate tool is installed.
+geoip2_geoipupdate_auto_update: true
+# The name to use for the SystemD service and timer that automatically updates
```
```suggestion
# The name to use for the systemd service and timer that automatically updates
```

jsf9k commented on tasks/main.yml:
```diff
@@ -23,3 +23,56 @@
+    - name: Set up automatic updates for GeoIP databases
+      when: geoip2_geoipupdate_auto_update
+      block:
+        - name: Write the geoipupdate SystemD service file
```
```suggestion
        - name: Write the geoipupdate systemd service file
```

jsf9k commented on tasks/main.yml:
```diff
@@ -23,3 +23,56 @@
+            mode: 0644
+            src: geoipupdate.service.j2
+
+        - name: Write the geoipupdate SystemD timer file
```
```suggestion
        - name: Write the geoipupdate systemd timer file
```

jsf9k commented on tasks/main.yml:
```diff
@@ -23,3 +23,56 @@
+            mode: 0644
+            src: geoipupdate.timer.j2
+
+        - name: Start and enable the geoipupdate SystemD timer
```
```suggestion
        - name: Start and enable the geoipupdate systemd timer
```

jsf9k reviewed (changes requested):
I have a minor quibble over the capitalization of "systemd"[^1], but otherwise this looks like strong work to me.

[^1]: I know I have capitalized it as "SystemD" in the past but I now know this to be wrong.

dav3r commented on README.md:
```diff
@@ -64,6 +66,12 @@ None.
+| geoip2\_geoipupdate\_auto\_update | Whether to configure automatic updates when `geoipupdate` is installed. | `true` | No |
+| geoip2\_geoipupdate\_service\_name | The name to use for the `geoipupdate` SystemD service and timer. | `geoipupdate` | No |
+| geoip2\_geoipupdate\_service\_timer\_on\_calendar | The calendar expression to use for the `geoipupdate` SystemD timer. | `Wed,Sat America/New_York` | No |
+| geoip2\_geoipupdate\_service\_timer\_randomized\_delay\_sec | The randomized delay in seconds to use for the `geoipupdate` SystemD timer. | `3h` | No |
```
I'm confused about the description here talking about the value being in "seconds", but the default value is `3h`.  Shouldn't that be `10800` seconds?

mcdonnnj replied:
Please let me know if [9c1a1eb](https://github.com/cisagov/ansible-role-geoip2/pull/51/commits/9c1a1eb0d15c22bfd6ee139b38d48dc7d715efdf) clarifies things.

dav3r replied:
It's better, but please see https://github.com/cisagov/ansible-role-geoip2/pull/51#discussion_r1664224157.

mcdonnnj replied:
Please see [450b394](https://github.com/cisagov/ansible-role-geoip2/pull/51/commits/450b394807a674789207cfc2951c2d16a59c19e1). I do not want to burn time trying to document the peculiarities of systemd's functionality, so I would rather just reference the options being populated and leave it to the reader.

dav3r replied:
This works for me- I'm not dying on this hill.

dav3r commented on defaults/main.yml:
```diff
@@ -10,3 +10,26 @@ geoip2_maxmind_editions:
+# Therefore, the default is to update the databases on Wednesdays and Saturdays
+# between 00:00 and 03:00 EST.
+geoip2_geoipupdate_service_timer_on_calendar: Wed,Sat America/New_York
+geoip2_geoipupdate_service_timer_randomized_delay_sec: 3h
```
Same confusion here as above re: `3h` vs. seconds.

mcdonnnj replied:
Please let me know if [9c1a1eb](https://github.com/cisagov/ansible-role-geoip2/pull/51/commits/9c1a1eb0d15c22bfd6ee139b38d48dc7d715efdf) clarifies things.

dav3r replied:
I'm in agreement with @jsf9k's suggestion [here](https://github.com/cisagov/ansible-role-geoip2/pull/51/files#r1664226422).

mcdonnnj replied:
Please see [450b394](https://github.com/cisagov/ansible-role-geoip2/pull/51/commits/450b394807a674789207cfc2951c2d16a59c19e1). I do not want to burn time trying to document the peculiarities of systemd's functionality, so I would rather just reference the options being populated and leave it to the reader.

dav3r replied:
I'm on board with the updated variable description, but I think this comment should remain where you had it in `defaults/main.yml`:
```
# This value must be a valid systemd.time time span value. Please see the
# documentation for more information:
# https://www.freedesktop.org/software/systemd/man/latest/systemd.time.html
```

mcdonnnj replied:
Ok. I only removed it because it was my own interpretation after researching. The [option's documentation](https://www.freedesktop.org/software/systemd/man/latest/systemd.timer.html#RandomizedDelaySec=) is not super helpful. Please see [3d9648a](https://github.com/cisagov/ansible-role-geoip2/pull/51/commits/3d9648aef4e4630e07cca31cb320913fafb566f7).

mcdonnnj replied:
@dav3r Any other concerns on this suggestion?

dav3r replied:
Nope, all good - resolving now!

dav3r reviewed (approved):
LGTM, though I noted one question. 👍

mcdonnnj committed:
Use the correct capitalization for systemd

In the words of the project: "Yes, it is written systemd, not system D
or System D, or even SystemD. And it isn't system d either."

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

mcdonnnj committed:
Improve documentation around a variable

Better explain what values can be provided for the the
geoip2_geoipupdate_service_timer_randomized_delay_sec variable.

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

jsf9k commented on README.md:
```diff
@@ -67,9 +67,9 @@ None.
-| geoip2\_geoipupdate\_service\_timer\_randomized\_delay\_sec | The randomized delay in seconds to use for the `geoipupdate` SystemD timer. | `3h` | No |
+| geoip2\_geoipupdate\_service\_name | The name to use for the `geoipupdate` systemd service and timer. | `geoipupdate` | No |
+| geoip2\_geoipupdate\_service\_timer\_on\_calendar | The calendar expression to use for the `geoipupdate` systemd timer. | `Wed,Sat America/New_York` | No |
+| geoip2\_geoipupdate\_service\_timer\_randomized\_delay\_sec | The `systemd.time` time span value that represents the randomized delay in seconds to use for the `geoipupdate` systemd timer. | `3h` | No |
```
I still feel like the mention of seconds is very misleading here:
```suggestion
| geoip2\_geoipupdate\_service\_timer\_randomized\_delay | The `systemd.time` time span value that represents the randomized delay to use for the `geoipupdate` systemd timer. | `3h` | No |
```

dav3r replied:
I agree with @jsf9k here.  If you really want to, you could add a clause along the lines of "if there's no time unit, then it means seconds", but I don't think it's totally necessary.

mcdonnnj replied:
Please see [450b394](https://github.com/cisagov/ansible-role-geoip2/pull/51/commits/450b394807a674789207cfc2951c2d16a59c19e1). I do not want to burn time trying to document the peculiarities of systemd's functionality, so I would rather just reference the options being populated and leave it to the reader.

jsf9k commented on defaults/main.yml:
```diff
@@ -32,4 +32,7 @@ geoip2_geoipupdate_service_name: geoipupdate
+# This value must be a valid systemd.time time span value. Please see the
+# documentation for more information:
+# https://www.freedesktop.org/software/systemd/man/latest/systemd.time.html
 geoip2_geoipupdate_service_timer_randomized_delay_sec: 3h
```
See [my related comment in `README.md`](https://github.com/cisagov/ansible-role-geoip2/pull/51#discussion_r1664224157):
```suggestion
geoip2_geoipupdate_service_timer_randomized_delay: 3h
```

mcdonnnj replied:
I do not want to change this because the name signifies that it is for the `RandomizedDelaySec` option.

jsf9k reviewed (commented):
I'm still perturbed over the seconds thing that @dav3r pointed out.

jsf9k commented on templates/geoipupdate.timer.j2:
```diff
@@ -0,0 +1,10 @@
+
+[Timer]
+OnCalendar={{ geoip2_geoipupdate_service_timer_on_calendar }}
+RandomizedDelaySec={{ geoip2_geoipupdate_service_timer_randomized_delay_sec }}
```
Please see [my related comment in `README.md`](https://github.com/cisagov/ansible-role-geoip2/pull/51#discussion_r1664224157):
```suggestion
RandomizedDelaySec={{ geoip2_geoipupdate_service_timer_randomized_delay }}
```

mcdonnnj replied:
Same comment about the variable name.

mcdonnnj committed:
Change how two variables are documented

Instead of trying to document the peculiarities of systemd, the
descriptions for the two variables used to modify the geoipupdate
timer are changed to simply refer to the options whose values are
populated.

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>
Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

mcdonnnj committed:
Restore a deleted comment

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

jsf9k reviewed (approved):
I'm more than happy with the new descriptions.
